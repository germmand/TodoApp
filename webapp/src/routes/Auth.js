import AuthViews from '../views/Auth';

const authRoutes = [
  {
    path: '/auth/login',
    component: AuthViews.Login,
    key: 'loginView',
  },
  {
    path: '/auth/signup',
    component: AuthViews.Signup,
    key: 'signupView',
  },
  {
    redirect: true,
    path: '/auth',
    to: '/auth/login',
    key: 'loginRedirect',
  },
];

export default authRoutes;
