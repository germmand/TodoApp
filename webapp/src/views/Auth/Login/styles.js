export default theme => ({
  root: {
    ...theme.mixins.gutters(),
    paddingTop: theme.spacing.unit * 2,
    paddingBottom: theme.spacing.unit * 2,
  },
  textField: {
    width: '100%',
    '& + &': {
      marginTop: theme.spacing.unit * 2,
    },
  },
  fields: {
    marginTop: theme.spacing.unit * 2,
    width: 500,
  },
  fieldError: {
    color: theme.palette.danger.main,
    marginBottom: theme.spacing.unit * 2,
    marginTop: theme.spacing.unit,
  },
  content: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
  },
  signInButton: {
    marginTop: theme.spacing.unit * 2,
    width: '100%',
  },
  topDivider: {
    width: '75%',
  },
  progress: {
    display: 'block',
    marginTop: theme.spacing.unit * 2,
    marginLeft: 'auto',
    marginRight: 'auto',
  },
});
