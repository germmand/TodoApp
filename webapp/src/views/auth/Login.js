import React from 'react';

const Login = () => (
    <div>
        <h2>Hey! you made it to login, broo! :)</h2>
    </div>
);

export default Login;
