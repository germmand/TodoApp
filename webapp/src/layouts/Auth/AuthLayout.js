// React
import React from 'react';
import PropTypes from 'prop-types';
import { Switch, Route, Redirect } from 'react-router-dom';

// Material-UI
import { withStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';

// Redux
import { connect } from 'react-redux';

// Auth Views
import authRoutes from '../../routes/Auth';

const authViews = (
    <Switch>
        {authRoutes.map((prop) => {
          if (prop.redirect) {
            return <Redirect from={prop.path} to={prop.to} key={prop.key} />;
          }
          return <Route path={prop.path} component={prop.component} key={prop.key} />;
        })}
    </Switch>
);

const styles = {
  grow: {
    flexGrow: 1,
  },
  viewContainer: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'column',
    height: '90vh',
  },
};

class AuthLayout extends React.Component {
    onHandleLoginRoute = () => {
      const { history, location } = this.props;
      if (location.pathname !== '/auth/login') {
        history.push('/auth/login');
      }
    };

    onHandleRegisterRoute = () => {
      const { history, location } = this.props;
      if (location.pathname !== '/auth/signup') {
        history.push('/auth/signup');
      }
    };

    onDashboardRedirect = () => {
      const { history, isLoggedIn } = this.props;
      if (isLoggedIn) {
        history.replace('/home/Dashboard');
      }
    }

    componentDidMount = () => {
      // Redirects to dashboard in case
      // the user is already logged in
      // on opening/rendering.
      this.onDashboardRedirect();
    }

    componentDidUpdate = () => {
      // Redirects to dashboard in case
      // the user is already logged in
      // on action.
      this.onDashboardRedirect();
    }

    render() {
      const { classes } = this.props;

      return (
        <div>
            <AppBar position="static">
                <Toolbar>
                    <Typography variant="h6" color="inherit" className={classes.grow}>
                        TodoApp - Welcome!
                    </Typography>
                    <Button variant="outlined" color="inherit" onClick={this.onHandleLoginRoute}>
                        Login
                    </Button>
                    <Button color="inherit" onClick={this.onHandleRegisterRoute}>
                        Register
                    </Button>
                </Toolbar>
            </AppBar>
            <div className={classes.viewContainer}>
                {authViews}
            </div>
        </div>
      );
    }
}

AuthLayout.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  location: PropTypes.objectOf(PropTypes.string).isRequired,
  history: PropTypes.shape({
    action: PropTypes.string.isRequired,
    block: PropTypes.func.isRequired,
    createHref: PropTypes.func.isRequired,
    go: PropTypes.func.isRequired,
    goBack: PropTypes.func.isRequired,
    goForward: PropTypes.func.isRequired,
    length: PropTypes.number.isRequired,
    listen: PropTypes.func.isRequired,
    location: PropTypes.objectOf(PropTypes.string).isRequired,
    push: PropTypes.func.isRequired,
    replace: PropTypes.func.isRequired,
  }).isRequired,
  isLoggedIn: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  isLoggedIn: state.authenticationReducer.isLoggedIn,
});

export default connect(
  mapStateToProps,
  null,
)(withStyles(styles)(AuthLayout));
