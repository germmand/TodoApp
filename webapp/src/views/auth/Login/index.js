import React from 'react';
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';
import {
  Paper,
  TextField,
  Button,
  Typography,
  Divider,
  CircularProgress,
} from '@material-ui/core';

import styles from './styles';

class Login extends React.Component {
    state = {
      values: {
        username: '',
        password: '',
      },
      isLoading: false,
    };

    onChangeUsername = (event) => {
      const { values } = this.state;
      const { value: username } = event.target;
      this.setState({
        values: { ...values, username },
      });
    };

    onChangePassword = (event) => {
      const { values } = this.state;
      const { value: password } = event.target;
      this.setState({
        values: { ...values, password },
      });
    };

    onHandleSignIn = () => {
      // This is temporary
      // Just to check that the CircularProgress is
      // being rendered properly.
      this.setState({
        isLoading: true,
      });
      setTimeout(() => {
        this.setState({
          isLoading: false,
        });
      }, 1500);
    };

    render() {
      const { classes } = this.props;
      const { values, isLoading } = this.state;

      return (
            <Paper className={classes.root} elevation={1}>
                <form className={classes.content}>
                    <Typography variant="h5" component="h3">
                        Welcome back!
                    </Typography>
                    <Divider className={classes.topDivider} />
                    <div className={classes.fields}>
                        <TextField
                            className={classes.textField}
                            label="Username"
                            name="username"
                            type="text"
                            variant="outlined"
                            value={values.username}
                            onChange={this.onChangeUsername}
                        />
                        <TextField
                            className={classes.textField}
                            label="Password"
                            name="password"
                            type="password"
                            variant="outlined"
                            value={values.password}
                            onChange={this.onchangePassword}
                        />
                    </div>
                    {isLoading ? (
                        <CircularProgress className={classes.progress} />
                    ) : (
                        <Button
                            className={classes.signInButton}
                            color="primary"
                            size="large"
                            variant="contained"
                            onClick={this.onHandleSignIn}
                        >
                            Sign in now!
                        </Button>
                    )}
                </form>
            </Paper>
      );
    }
}

Login.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
};

export default withStyles(styles)(Login);
