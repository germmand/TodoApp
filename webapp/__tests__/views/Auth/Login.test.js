import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import debounce from 'lodash/debounce';

import { RawLogin } from '../../../src/views/Auth/Login';

jest.mock('lodash/debounce');

describe('Login component', () => {
  beforeEach(() => {
    debounce.mockImplementation(fn => fn);
  });

  it('matches snapshot', () => {
    const props = {
      classes: {
        container: 'container-class',
        root: 'root-class',
        content: 'content-class',
        topDivider: 'topDivider-class',
        fields: 'fields-class',
        textField: 'textField-class',
        signInButton: 'signInButton-class',
      },
      isLogginIn: false,
      login: jest.fn(),
    };
    const LoginComponent = renderer.create(<RawLogin {...props} />);
    expect(LoginComponent.toJSON()).toMatchSnapshot();
  });

  it('calls login action creator on sign in', () => {
    const props = {
      classes: {
        signInButton: 'signInButton-class',
      },
      isLogginIn: false,
      authError: {
        username: ['This field can\'t be blank.'],
      },
      login: jest.fn(() => Promise.resolve({})),
    };
    const credentials = {
      username: 'test_user',
      password: 'test_password',
    };
    const LoginComponent = mount(<RawLogin {...props} />).find('Login');
    LoginComponent.setState({ values: credentials }, () => {
      const LoginButton = LoginComponent.find('Button')
        .filterWhere(s => s.props()
          .className
          .includes(props.classes.signInButton));
      expect(LoginButton).toHaveLength(1);
      LoginButton.simulate('click');
      expect(props.login).toHaveBeenCalledWith(credentials);
    });
  });

  it('shows server auth errors is wasSubmitted is true', () => {
    const props = {
      classes: { },
      isLogginIn: false,
      authError: {
        username: ['This field can\'t be blank.'],
      },
      login: jest.fn(() => Promise.resolve({})),
    };
    let LoginComponent = mount(<RawLogin {...props} />).find('Login');
    LoginComponent.setState({ wasSubmitted: true }, () => {
      LoginComponent = LoginComponent.update().find('Login');
      const ServerAuthErrorsComponent = LoginComponent.find('ServerAuthErrors');
      expect(ServerAuthErrorsComponent).toHaveLength(1);
      expect(ServerAuthErrorsComponent.props().errors).toEqual(props.authError);
    });
  });

  it('sets wasSubmitted to true if sign in button is clicked', () => {
    const props = {
      classes: {
        signInButton: 'signInButton-class',
      },
      isLogginIn: false,
      authError: {
        username: ['This field can\'t be blank.'],
      },
      login: jest.fn(() => Promise.resolve({})),
    };
    const LoginComponent = mount(<RawLogin {...props} />).find('Login');
    const LoginButton = LoginComponent.find('Button')
      .filterWhere(s => s.props()
        .className
        .includes(props.classes.signInButton));
    expect(LoginButton).toHaveLength(1);
    LoginButton.find('button').simulate('click');
    setImmediate(() => {
      expect(LoginComponent.state().wasSubmitted).toBeTruthy();
    });
  });

  it('username is updated successfully on state when change event occurs', () => {
    const props = {
      classes: { },
      isLogginIn: false,
      login: jest.fn(() => Promise.resolve({})),
    };
    const username = 'test_user';
    const LoginComponent = mount(<RawLogin {...props} />).find('Login');
    const UsernameInput = LoginComponent.find('TextField')
      .filterWhere(s => s.props()
        .name === 'username');
    expect(UsernameInput).toHaveLength(1);
    UsernameInput.props().onChange({ target: { value: username } });
    expect(LoginComponent.state().values.username).toEqual(username);
    expect(LoginComponent.state().touched.username).toBeTruthy();
    expect(LoginComponent.state().errors.username).toEqual(undefined);
  });

  it('username errors are placed in state correctly when errors on username input', () => {
    const props = {
      classes: { },
      isLogginIn: false,
      login: jest.fn(() => Promise.resolve({})),
    };
    const username = '';
    const errorsExpected = ['Username is required'];
    const LoginComponent = mount(<RawLogin {...props} />).find('Login');
    const UsernameInput = LoginComponent.find('TextField')
      .filterWhere(s => s.props()
        .name === 'username');
    expect(UsernameInput).toHaveLength(1);
    UsernameInput.props().onChange({ target: { value: username } });
    expect(LoginComponent.state().values.username).toEqual(username);
    expect(LoginComponent.state().touched.username).toBeTruthy();
    expect(LoginComponent.state().errors.username).toEqual(errorsExpected);
  });

  it('password is updated successfully on state when change event occurs', () => {
    const props = {
      classes: { },
      isLogginIn: false,
      login: jest.fn(() => Promise.resolve({})),
    };
    const password = 'super_password';
    const LoginComponent = mount(<RawLogin {...props} />).find('Login');
    const PasswordInput = LoginComponent.find('TextField')
      .filterWhere(s => s.props()
        .name === 'password');
    expect(PasswordInput).toHaveLength(1);
    PasswordInput.props().onChange({ target: { value: password } });
    expect(LoginComponent.state().values.password).toEqual(password);
    expect(LoginComponent.state().touched.password).toBeTruthy();
    expect(LoginComponent.state().errors.password).toEqual(undefined);
  });

  it('password errors are placed in state correctly when errors on password input', () => {
    const props = {
      classes: { },
      isLogginIn: false,
      login: jest.fn(() => Promise.resolve({})),
    };
    const password = '';
    const errorsExpected = ['Password is required'];
    const LoginComponent = mount(<RawLogin {...props} />).find('Login');
    const PasswordInput = LoginComponent.find('TextField')
      .filterWhere(s => s.props()
        .name === 'password');
    expect(PasswordInput).toHaveLength(1);
    PasswordInput.props().onChange({ target: { value: password } });
    expect(LoginComponent.state().values.password).toEqual(password);
    expect(LoginComponent.state().touched.password).toBeTruthy();
    expect(LoginComponent.state().errors.password).toEqual(errorsExpected);
  });

  it('renders client-side username errors', () => {
    const props = {
      classes: { },
      isLogginIn: false,
      login: jest.fn(() => Promise.resolve({})),
    };
    const state = {
      touched: {
        username: true,
      },
      errors: {
        username: ['Username is required.'],
      },
    };
    let LoginComponent = mount(<RawLogin {...props} />).find('Login');
    LoginComponent.setState({ ...state }, () => {
      LoginComponent = LoginComponent.update().find('Login');
      const UsernameTypographyError = LoginComponent.find('Typography')
        .filterWhere(s => s.props().id === 'usernameFieldError');
      expect(UsernameTypographyError).toHaveLength(1);
      expect(UsernameTypographyError.text()).toEqual(state.errors.username[0]);
    });
  });

  it('renders client-side password errors', () => {
    const props = {
      classes: { },
      isLogginIn: false,
      login: jest.fn(() => Promise.resolve({})),
    };
    const state = {
      touched: {
        password: true,
      },
      errors: {
        password: ['Password is required.'],
      },
    };
    let LoginComponent = mount(<RawLogin {...props} />).find('Login');
    LoginComponent.setState({ ...state }, () => {
      LoginComponent = LoginComponent.update().find('Login');
      const PasswordTypographyError = LoginComponent.find('Typography')
        .filterWhere(s => s.props().id === 'passwordFieldError');
      expect(PasswordTypographyError).toHaveLength(1);
      expect(PasswordTypographyError.text()).toEqual(state.errors.password[0]);
    });
  });

  it('replaces Button to CircularProgress when isLogginIn is true', () => {
    const props = {
      classes: {
        signInButton: 'signInButton-class',
        progress: 'progress-class',
      },
      isLogginIn: true,
      login: jest.fn(() => Promise.resolve({})),
    };
    const LoginComponent = mount(<RawLogin {...props} />).find('Login');
    const CircularProgress = LoginComponent.find('CircularProgress')
      .filterWhere(s => s.props().className.includes(props.classes.progress));
    const LoginButton = LoginComponent.find('Button')
      .filterWhere(s => s.props().className.includes(props.classes.signInButton));
    expect(CircularProgress).toHaveLength(1);
    expect(LoginButton).toHaveLength(0);
  });
});
