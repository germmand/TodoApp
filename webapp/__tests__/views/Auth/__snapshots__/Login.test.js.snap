// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component matches snapshot 1`] = `
<div
  className="Login-container-2 container-class"
>
  <div
    className="MuiPaper-root-10 MuiPaper-elevation1-13 MuiPaper-rounded-11 Login-root-1 root-class"
  >
    <form
      className="Login-content-3 content-class"
    >
      <h3
        className="MuiTypography-root-37 MuiTypography-h5-53"
      >
        Welcome back!
      </h3>
      <hr
        className="MuiDivider-root-73 Login-topDivider-8 topDivider-class"
      />
      <div
        className="Login-fields-5 fields-class"
      >
        <div
          className="MuiFormControl-root-78 Login-textField-4 textField-class"
        >
          <label
            className="MuiFormLabel-root-93 MuiInputLabel-root-82 MuiInputLabel-formControl-87 MuiInputLabel-animated-90 MuiInputLabel-outlined-92"
            data-shrink={false}
          >
            Username
          </label>
          <div
            className="MuiInputBase-root-113 MuiOutlinedInput-root-100 MuiInputBase-formControl-114"
            onClick={[Function]}
          >
            <fieldset
              aria-hidden={true}
              className="MuiPrivateNotchedOutline-root-130 MuiOutlinedInput-notchedOutline-107"
              style={
                Object {
                  "paddingLeft": 8,
                }
              }
            >
              <legend
                className="MuiPrivateNotchedOutline-legend-131"
                style={
                  Object {
                    "width": 0.01,
                  }
                }
              >
                <span
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "&#8203;",
                    }
                  }
                />
              </legend>
            </fieldset>
            <input
              aria-invalid={false}
              className="MuiInputBase-input-123 MuiOutlinedInput-input-108"
              disabled={false}
              name="username"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="MuiFormControl-root-78 Login-textField-4 textField-class"
        >
          <label
            className="MuiFormLabel-root-93 MuiInputLabel-root-82 MuiInputLabel-formControl-87 MuiInputLabel-animated-90 MuiInputLabel-outlined-92"
            data-shrink={false}
          >
            Password
          </label>
          <div
            className="MuiInputBase-root-113 MuiOutlinedInput-root-100 MuiInputBase-formControl-114"
            onClick={[Function]}
          >
            <fieldset
              aria-hidden={true}
              className="MuiPrivateNotchedOutline-root-130 MuiOutlinedInput-notchedOutline-107"
              style={
                Object {
                  "paddingLeft": 8,
                }
              }
            >
              <legend
                className="MuiPrivateNotchedOutline-legend-131"
                style={
                  Object {
                    "width": 0.01,
                  }
                }
              >
                <span
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "&#8203;",
                    }
                  }
                />
              </legend>
            </fieldset>
            <input
              aria-invalid={false}
              className="MuiInputBase-input-123 MuiOutlinedInput-input-108 MuiInputBase-inputType-126"
              disabled={false}
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="password"
              value=""
            />
          </div>
        </div>
      </div>
      <button
        className="MuiButtonBase-root-158 MuiButton-root-132 MuiButton-contained-143 MuiButton-containedPrimary-144 MuiButton-raised-146 MuiButton-raisedPrimary-147 MuiButton-sizeLarge-156 Login-signInButton-7 signInButton-class"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-133"
        >
          Sign in now!
        </span>
        <span
          className="MuiTouchRipple-root-161"
        />
      </button>
    </form>
  </div>
</div>
`;
